substratevm/src/com.oracle.svm.agent/src/com/oracle/svm/agent/BreakpointInterceptor.java

```diff
@@ -678,6 +699,124 @@ private static boolean handleGetSystemResources(JNIEnvironment jni, Breakpoint b
+     * parameter is null.
+     */
+    @SuppressWarnings("unused")
+    private static boolean postDefineClass(JNIEnvironment jni, Breakpoint bp) {
+        boolean isDynamicallyGenerated = false;
+
+        // Get class name from the argument "name" of
+        // defineClass(String name, byte[] b, int off, int len)
+        // The first argument is implicitly "this", so "name" is the 2nd parameter.
+        String nameFromDefineClassParam = fromJniString(jni, getObjectArgument(1, 1));
+        final String definedClassName;
+        JNIObjectHandle self = getObjectArgument(0);
+        // 1. Don't have a name for class before defining.
+        // The class is dynamically generated.
+        if (nameFromDefineClassParam == null) {
+            isDynamicallyGenerated = true;
+            // Get name from parameter "c" of method postDefineClass(Class<?> c, ProtectionDomain
+            // pd)
+            definedClassName = getClassNameOrNull(jni, getObjectArgument(1));
+        } else {
+            definedClassName = nameFromDefineClassParam;
+            // Filter out internal classes which are definitely not dynamically generated
+            // CallerClass is always java.lang.ClassLoader, we only check the defined class
+            AccessAdvisor postDefineCLassAccessAdvisor = new AccessAdvisor();
+            postDefineCLassAccessAdvisor.setInLivePhase(true);
+            if (postDefineCLassAccessAdvisor.shouldIgnore(new LazyValue<>(() -> definedClassName), new LazyValue<>(() -> null))) {
+                isDynamicallyGenerated = false;
+            }
+
+            // 2. Class with name starts with $ or contains $$ is usually dynamically generated
+            String className = definedClassName.substring(definedClassName.lastIndexOf('.') + 1);
+            if (className.startsWith("$") || className.contains("$$")) {
+                isDynamicallyGenerated = true;
+            } else {
+                // 3. A dynamically defined class always return null
+                // when call java.lang.ClassLoader.getResource(classname)
+                // This is the accurate but slow way.
+                String asResourceName = definedClassName.replace('.', '/') + ".class";
+                try (CCharPointerHolder resourceNameHolder = toCString(asResourceName);) {
+                    JNIObjectHandle resourceNameJString = jniFunctions().getNewStringUTF().invoke(jni, resourceNameHolder.get());
+                    JNIObjectHandle returnValue = callObjectMethodL(jni, self, agent.handles().javaLangClassLoaderGetResource, resourceNameJString);
+                    isDynamicallyGenerated = returnValue.equal(nullHandle());
+                }
+            }
+        }
+
+        // Reuse verifyForName
+        // CallerClass is always java.lang.ClassLoader, we only check the defined class
+        boolean allowed = (accessVerifier == null || accessVerifier.verifyForName(jni, nullHandle(), definedClassName));
```

Comment:
In the new version of the code we need to pass an extra argument here. I am working with @d-kozak to see how to do that best.

Reply:
Fixed in my branch.